pkg/corerp/frontend/controller/applications/graph_util.go

```diff
@@ -0,0 +1,577 @@
+//
+// This function does not return errors and will ignore missing or corrupted data. It is expected that the caller
+// will display the results to a human user, so rather than failing to computeGraph the graph, we will return partial
+// results. Each ApplicationGraphResource will have a provisioning state that indicates whether the resource
+// was successfully processed or not.
+func computeGraph(applicationName string, applicationResources []generated.GenericResource, environmentResources []generated.GenericResource) *corerpv20231001preview.ApplicationGraphResponse {
+	// The first step is to figure out what resources are part of the application.
+	//
+	// Since Radius has both application-scoped and environment-scoped resources we need to merge the two lists
+	// and take care to annotate each resource with whether it's part of the application or not. Any application-scoped
+	// resource will be part of the environment, so we also filter duplicates.
+	//
+	// We produce two data-structures:
+	// - resources (all "known" resources) using the API wire format
+	// - resourcesByIDInApplication (a map of resource ID to whether the resource is part of the application).
+	//
+	// This allows us to reason about three sets of resources:
+	// - Resources that belong to the application: resourcesByIDInApplication[id] == true
+	// - Resources that are "known" but don't belong to the application: resourcesByIDInApplication[id] == false
+	// - Resources that "external":  ok, _ := resourcesByIDInApplication[id]; !ok (not in the map)
+	resources := []generated.GenericResource{}
+	resourcesByIDInApplication := map[string]bool{}
+
+	for _, resource := range applicationResources {
+		resources = append(resources, resource)
+
+		// Application-scoped resources are by definition "in" the application
+		resourcesByIDInApplication[*resource.ID] = true
+	}
+
+	for _, resource := range environmentResources {
+		_, found := resourcesByIDInApplication[*resource.ID]
+		if found {
+			// Appears in both application and environment lists, avoid duplicates.
+			continue
+		}
+
+		// This is an environment-scoped resource. We need to process the connections
+		// to determine if it's part of the application.
+		resources = append(resources, resource)
+		resourcesByIDInApplication[*resource.ID] = false
+	}
+
+	// Next we need to process each entry in the resources list and build up the application graph.
+	applicationGraphResourcesByID := map[string]corerpv20231001preview.ApplicationGraphResource{}
+	for _, resource := range resources {
+		applicationGraphResource := applicationGraphResourceFromID(*resource.ID)
+		if applicationGraphResource == nil {
+			continue // Invalid resource ID, skip
+		}
+
+		//update provisioning state
+		provisioningState, ok := resource.Properties["provisioningState"]
+		if ok {
+			// ignore err since in that case we just default to "succeeded"
+			state, _ := provisioningState.(string)
```

Comment:
Since the func applicationGraphResourceFromID() constructs the applicationGraphResource witht  provisioningState set to success. We default to this, since only for Async, we have the field and the state could be different. Otherwise we consider it as a success ( since sync responses would have indicated otherwise immediately)